### DISCUSSION TITLE
HDR Playback Support

### DISCUSSION THREAD

**Discussion by qvattr0** [prayag17/Blink#212] [Q&A]
As far as I remember, the underlying video player of Jellyplayer is VLC, correct? If that's the case, does Jellyplayer support HDR playback? I've noticed some very bizarre video playback/rendering issues (video being very dark when UI is hidden, and blown out when UI is present) so I was wondering if the issues I'm facing are exclusive to my machine. I've just tested and the issue is present with regular non-HDR videos too: the image gets darker (the dark segments, at least) when the UI is hidden; and once the UI is revealed again via a mouse movement or something similar, the shadows get blown out.

**Comment by prayag17** [comment]
JellyPlayer does NOT use VLC, all video playback is handled by the respective native webviews.

The video brightness issue is not something I have noticed, can you tell me your OS and device?

**Reply by qvattr0** [reply]
Sure, my OS is Windows 11 2H22. The device is a PC (not sure if that's useful) driving a Samsung G8 OLED monitor. I've actually been trying to document the color issues and have collected some images that will be provided below:

Jellyplayer (visible UI):
![Jellyplayer (Visible UI)](https://github.com/prayag17/JellyPlayer/assets/141742777/6d55dc70-59d9-4115-910c-3efea2d6cc9f)

Jellyplayer (hidden UI):
![Jellyplayer (Hidden UI)](https://github.com/prayag17/JellyPlayer/assets/141742777/a4ede741-9b56-4bd5-a886-096606a1e967)

VLC4 (true image):
![VLC4 (control)](https://github.com/prayag17/JellyPlayer/assets/141742777/8ab4bc85-ac4d-40e2-9fe1-3158b0d6883a)

The difference is surprisingly distinguishable even in SDR mode. Let me know if you'd like me to provide you with any other information.

**Reply by qvattr0** [reply]
Also, what do you mean by respective "native webviews?" I'm not particularly acquainted with the world of video playback in Windows (or at least under-the-hood playback).

**Reply by prayag17** [reply]
By native Webviews I mean that the renderer for the app using web pages

**Reply by qvattr0** [reply]
Understood, will have to do some research to see if I can resolve the issue myself. What do you think though? About the color reproduction/brightness.

**Reply by qvattr0** [reply]
I have discovered another tiny detail of this issue: apparently, when in SDR mode, the image (a non-HDR video file) returns to proper color representation a second or two after OSD disappears (normal color representation while OSD is up -> OSD disappears -> the color representation of the image is darker than expected -> 1-2 seconds pass -> color representation returns to normal). Given this new information, I'd suppose the root of the issue is potentially associated with any image effects the OSD overlay has when it appears/disappears.

**Reply by GrabbenD** [reply]
MPV is known to solve all issues related to transcoding and should be less of an headache to work with (in comparson to using Media3 or LibVLC)

Not sure how hard it would be to wire it into the existing `react-player` setup but this might be worth looking into since they have React support: https://github.com/Kagami/mpv.js

**Reply by qvattr0** [reply]
> MPV is known to solve all issues related to transcoding and should be less of an headache to work with (in comparson to using Media3 or LibVLC)
> 
> Not sure how hard it would be to wire it into the existing `react-player` setup but this might be worth looking into since they have React support: https://github.com/Kagami/mpv.js

Thanks for the suggestion! I was wondering if it would be realistic to implement MPV into Jellyplayer and wasn't aware of the existence of an MPV build that supported React. If this is never resolved, I'll consider putting in some of my own time to add MPV support.

**Reply by prayag17** [reply]
MPV.js is not maintained and has tons of issues. Implementing mpv with tauri is a bit difficult since there are no maintained mpv libs for rust, even VLC which has an official rust library is not being maintained. And tauri is still pretty new so there is not a way to implement a native canvas and overlay a webview renderer for ui. 
Though I do plan to add gstreamer which has excellent rust or create my own mpv library support once tauri adds native canvas render support. 

Note: I do know wry supports overlaying multiple windows but I have found no way to use it with tauri.

Also MPV.js is for electron and wont work in this app

**Reply by mitzsch** [reply]
Just want to peek into this discussion (although this feels like cheating on Plex as I am a real Plex advocate... whatever, time might have come to explore alternatives...)

It would be awesome if **mpv** could be incorporated into the player, for numerous reasons:
- proper audio support => audio passthrough, bitperfect audio output
- proper video playback support => **HDR tone mapping, HDR passthrough**, gpu-next/libplacebo...
- proper subtitle support => pgs, ass, srt you name it...
- highly customizable => create a mpv.conf and put all your necessary commands in there...
- change mpv.dll as needed
- future-proofing
- don´t rely on codecs supported by the underlying framework - mpv can play them all (almost...)
- many other reasons...


mpv can also be embedded via its --wid parameter. (https://mpv.io/manual/master/#options-wid). Maybe this way its "easier" to be implemented? This method is working great - for example, apps like Plex HTPC also rely on this parameter...


If the Jellyfin project would have a real HTPC companion, with a nice UI and all the features that real home theater enthusiasts need, it would be the preferred choice over plex. The Jellyfin Media Player is just based on the EOL PlexMediaPlayer and is already in a dead spot - no gpu-next support possible... and other alternatives like emby home theater.... we don´t need to talk about this one... (internet access is needed for just showing the ui?!)

**Reply by GrabbenD** [reply]
@mitzsch Has great points. Instead of dealing with media players which have a varying degree of supported video- and audio codecs, it'd be more robust if the backend was constantly improved upon through a upstream FOSS codebase like MPV. Here's examples of projects doing this already:
* The new and shiny [Jellyfin Media Player](https://github.com/jellyfin/jellyfin-media-player) desktop client is built for MPV backend. Additionally the Jellyfin team is slowly laying the [groundwork](https://github.com/jellyfin/jellyfin-androidtv/pull/3039) for Android TV to support newer and more feature rich backends like MPV.
* Another big client which provides MPV backend support is [Findroid](https://github.com/jarnedemeulemeester/findroid#features).

However, it's worth to keep in mind what @prayag17 already pointed out; realistically this might not be a feasible for a long time since there's no straightforward way of adding MPV into Tauri which JellyPlayer is built upon..

**We'd have to attract some knowledgeable developers from the Rust community to make it happen.**

**Reply by mitzsch** [reply]
> The new and shiny [Jellyfin Media Player](https://github.com/jellyfin/jellyfin-media-player) desktop client is built for MPV backend.

Well, as said the Jellyfin Media Player is based on the abandoned Plex Media Player. Its base does not allow to implement mpv in a way that we can get all the newest features from it - so the app is basically dead on arrival...  Have a look at this issue => https://github.com/jellyfin/jellyfin-media-player/issues/523 - That was one of the reasons why Plex HTPC was introduced... Its a recode that allows "full" mpv support. 

It does so by using the --wid parameter. Maybe this is a possible way to get mpv also into this player here? With this method, mpv creates its own windows and the app using it only needs to embed the whole window. Would something like this be possible with Tauri? @prayag17?

> We'd have to attract some knowledgeable developers from the Rust community to make it happen.

Absolutely!!!


(On a side note there was also a project that used mpv as its engine... => https://www.reddit.com/r/jellyfin/comments/ekmsz9/jellyfin_desktop_player_wip/ It seems dead... Maybe I will write him a message. Maybe he could release the code...)

**Reply by prayag17** [reply]
@mitzsch I can't promise an mpv player implementation yet as I mentioned before cause there is no maintained MPV rs library available to use so the only way to implement mpv player would be to create an mpv library(I don't really have time to do that). 
But I do have plans for implementing Gstreamer but that's not gonna happen anytime soon.
Also, currently there no straight forward way for implementing a native canvas for rendering video frames with the help of GPU tho the Tauri team is working on it. But I can't guarantee an MPV or Gstreamer or VLC implementation for v1 atleast 

> (On a side note there was also a project that used mpv as its engine... => https://www.reddit.com/r/jellyfin/comments/ekmsz9/jellyfin_desktop_player_wip/ It seems dead... Maybe I will write him a message. Maybe he could release the code...

That client was actually the motivation for me to create JellyPlayer, I contacted the dev a long time ago but he didn't reply

**Reply by prayag17** [reply]
Update: with tauri V2 on its way with [multiwebview support](https://beta.tauri.app/blog/tauri-2-0-0-beta/#multiwebview), I might be able to get [gstreamer](https://gstreamer.freedesktop.org/) to work with tauri but I cant promise a proper implementation right now. Just wanted to update this thread to current info.

Also, if you notice your video quality degrading while playing video on windows you should disable `Process video automatically to enhance it (depends on your device hardware)` option at `Apps -> Video Playback` and you should get all [Media Foundation codecs installed from microsoft store](https://www.codecguide.com/media_foundation_codecs.htm) to get best video playback support

**Reply by qvattr0** [reply]
This gstreamer thing sounds very promising! Having HDR support on JellyPlayer would basically turn the player into the all-in-one solution for viewing content! (without having to deal with setting up mpv, which can be rather annoying). Also, thanks for the tip; I'll try it out once I have the time and report back the results.
